bin/preprocess.ml

```diff
@@ -0,0 +1,82 @@
+open Stdune
+open Import
+
+let doc = "Print the preprocessed output"
+
+let man =
+  [ `S "DESCRIPTION"
+  ; `P
+      {|$(b,dune preprocess FILE) build the given FILE and print the preprocessed output|}
+  ; `Blocks Common.help_secs
+  ]
+
+let info = Term.info "preprocess" ~doc ~man
+
+let change_ext f path =
+  let base, ext = Path.of_string path |> Path.split_extension in
+  let suffix = f ext in
+  Path.extend_basename base ~suffix |> Path.to_string
+
+let pp_with_ocamlc sctx pp_ml =
+  let ocamlc = (Super_context.context sctx).ocamlc |> Path.to_string in
+  let env = Super_context.context_env sctx |> Env.to_unix |> Array.of_list in
+  let argv =
+    Array.of_list [ocamlc; "-stop-after"; "parsing"; "-dsource"; pp_ml; "-dump-into-file"]
+  in
+  let pid =
+    Unix.create_process_env ocamlc argv env Unix.stdin Unix.stdout Unix.stderr
+  in
+  match Unix.waitpid [] pid |> snd with
+  | Unix.WEXITED 0 ->
+    let dump_file =
+      pp_ml |> change_ext (function
+        | ".mli" -> ".cmi.dump"
+        | ".ml"  -> ".cmo.dump"
+        | _ -> failwith "unreachable")
+    in
+    if not (Path.of_string dump_file |> Path.is_file) then
+      User_error.raise [ Pp.textf "cannot find a dump file: %s" dump_file ]
+    else
+      In_channel.with_open_text dump_file (fun ic -> Io.copy_channels ic stdout);
+      Sys.remove dump_file;
+      exit 0
+  | Unix.WEXITED n -> exit n
+  | Unix.WSIGNALED _ | Unix.WSTOPPED _ -> exit 255
+
+let term =
+  let+ common = Common.term
+  and+ file = Arg.(value & pos 0 string "" & info [] ~docv:"FILE")
+  and+ ctx_name =
+    Common.context_arg ~doc:{|Select context where to build.|} in
+  let config = Common.init common in
+  let file = Common.prefix_target common file in
+  if String.is_empty file then
+    User_error.raise [ Pp.textf "no file is given" ];
+  let pp_file =
+    file |> change_ext (function
+      | ".mli" -> ".pp.mli"
+      | ".ml"  -> ".pp.ml"
```

Comment:
This doesn't take into account "dialects" which can modify the set of valid extensions for source files, see `Dialect.extension`.

Reply:
Fixed it so that it uses the current dialect to check if it is an interface or not.